src/utils/Exceptions.h

```diff
@@ -0,0 +1,107 @@
+    to std::assert in that file will still be governed by NDEBUG)
+
+    It also provides two wrappers on std::exception and its derived classes
+    (e.g., std::runtime_error, std::domain_error, etc.) The two wrappers are:
+    1. dftefe::utils::throwException(expr,msg): a generic exception handler
+    which throws an optional message (msg) if expr evaluates to false. It
+    combines std::exception with an  additional messgae. (Note: the
+    std::exception has no easy way of taking in a message).
+    2. dftefe::utils::throwException<T>(expr, msg): similar to the above, but
+    takes a specific derived class of std::exception handler as a template
+    parameter. The derived std::exception must have a constructor that takes in
+    a string. For the ease of the user, we have typedef-ed some commonly used
+    derived classes of std::exception. A user can use the typedefs as the
+    template parameter instead. Available typedefs LogicError - std::logic_error
+       InvalidArgument - std::invalid_argument
+       DomainError - std::domain_error
+       LengthError - std::length_error
+       OutOfRangeError - std::out_of_range
+       FutureError - std::future_error
+       RuntimeError	- std::runtime_error
+       OverflowError - std::overflow_error
+       UnderflowError - std::underflow_error
+    */
+
+#undef Assert
+#undef AssertWithMsg
+
+#if defined(DFTEFE_DISABLE_ASSERT) || \
+  (!defined(DFTEFE_ENABLE_ASSERT) && defined(NDEBUG))
+#  include <assert.h> // .h to support old libraries w/o <cassert> - effect is the same
+#  define Assert(expr) ((void)0)
+#  define AssertWithMsg(expr, msg) ((void)0)
+
+#elif defined(DFTEFE_ENABLE_ASSERT) && defined(NDEBUG)
+#  undef NDEBUG // disabling NDEBUG to forcibly enable assert for sources that
+                // set DFTEFE_ENABLE_ASSERT even when in release mode (with
+                // NDEBUG)
+#  include <assert.h> // .h to support old libraries w/o <cassert> - effect is the same
+#  define Assert(expr) assert(expr)
+#  define AssertWithMsg(expr, msg) assert((expr) && (msg))
+
+#else
+#  include <assert.h> // .h to support old libraries w/o <cassert> - effect is the same
+#  define Assert(expr) assert(expr)
+#  define AssertWithMsg(expr, msg) assert((expr) && (msg))
+
+#endif
+
+    typedef std::logic_error      LogicError;
+    typedef std::invalid_argument InvalidArgument;
+    typedef std::domain_error     DomainError;
+    typedef std::length_error     LengthError;
+    typedef std::out_of_range     OutOfRangeError;
+    typedef std::future_error     FutureError;
+    typedef std::runtime_error    RuntimeError;
+    typedef std::overflow_error   OverflowError;
+    typedef std::underflow_error  UnderflowError;
+
+    void
+    throwException(bool condition, std::string msg = "");
+
+    template <class T>
+    void
+    throwException(bool condition, std::string msg = "");
+
+  } // namespace utils
+} // namespace dftefe
+#include "Exceptions.t.cpp"
+#endif // dftefeAssert_h
```

Comment:
minor type: comment should be "dftefeExceptions_h"

Reply:
Done